### ISSUE TITLE
Use ? to lift missing values

### ISSUE THREAD

**Issue by juliohm** [JuliaLang/julia#36628]
As discussed in https://discourse.julialang.org/t/status-of-question-mark-syntax-for-missing-values/27189 it would be really nice to have this syntax at some future release where one could type `T?` to indicate `Union{T,Missing}`. Possibly, also write `mean?(xs)` to automatically handle inputs with missing values.

**Comment by bkamins** [comment]
I (or even we as people working with data 😄) might be opinionated, but after a longer time of working with both `nothing` and `missing`  I find `Union{T, Missing}` to be used every day, while `Union{T, Nothing}` is much less frequent.

Also an especially annoying thing with `Union{T, Missing}` is that you have to write e.g. `Vector{Union{T, Missing}}` all the time, while I assume that `Nothing` most often comes just as `Union{T, Nothing}` without being an indicator of element in the collection.

In summary - I support the proposal by @juliohm - if this is possible to extend a parser in this way of course (I have not analyzed the consequences in detail, but it seems safe as in `? : ` we require space before `?`).

**Comment by bkamins** [comment]
Also then I would rise again the idea of:

> Another use of ? syntax (if we decided to go for missing with it) is that then we could print `?` (possibly using a different color in REPL) instead of `missing` when we show tabular data (similarly to the change in the way we now show `Bool` in some cases).

**Comment by KristofferC** [comment]
What "lift missing values" is needs to be specified.

>  but after a longer time of working with both nothing and missing I find Union{T, Missing} to be used every day, while Union{T, Nothing} is much less frequent.

It's interesting because I use `Union{T, Nothing}` all the time, in all kinds of different orthogonal packages. Just as an example: https://github.com/KristofferC/TOMLX.jl/blob/master/src/parser.jl.

**Comment by DilumAluthge** [comment]
We could have syntax for both `Union{T, Missing}` and `Union{T, Nothing}`.

Maybe `T?` could be shorthand for `Union{T, Missing}`.

And `T!` could be shorthand for `Union{T, Nothing}`.

**Comment by tkf** [comment]
There are several RFCs for adding API based on `Union{Some{T}, Nothing}` (https://github.com/JuliaLang/julia/pull/34821, https://github.com/JuliaLang/julia/pull/33758).  For example, it'd be nice `dict?[key]` to return `Union{Some{T}, Nothing}` while it makes less sense to return `Union{T,Missing}` because `dict` may contain `missing`s.

> And `T!` could be shorthand for `Union{T, Nothing}`.

`T!` is already a valid name so we can't use it.

**Comment by DilumAluthge** [comment]
We could come up with another syntax. Maybe `T?` for missing and `T??` for nothing.

**Comment by vtjnash** [comment]
Should we make them the same, and simply change the definition to `const Missing = Nothing`?
[_(me) quickly ducks and runs away_]

I think it could be worthwhile to have a JuliaCon BoF topic to hammer this discussion out, since we can't rely on hallway conversations this year! And syntax would be really nice for this.

**Comment by DilumAluthge** [comment]
Isn't the math different?
```julia
julia> 1 + missing
missing

julia> 1 + nothing
ERROR: MethodError: no method matching +(::Int64, ::Nothing)
Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...) at operators.jl:538
  +(::T, ::T) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8} at int.jl:86
  +(::Union{Int16, Int32, Int64, Int8}, ::BigInt) at gmp.jl:531
  ...
Stacktrace:
 [1] top-level scope at REPL[2]:1
```

**Comment by tkf** [comment]
> Should we make them the same

Actually, I just made an exactly the same comment in Zulip:

> By the way, I think concrete differences between `missing` and `nothing` in Julia ATM is that `missing` has more defined behaviors like `isless(1, missing)` and `missing | true` (other than "manual lifting" like `missing + 1`).  Is there anything else?
>
> I get that `isless(1, missing)` is super useful. But I wonder if people heavily rely on the 3VL?  If not, maybe we can define `isless` for `nothing` and then merge the two kinds of missingness?
>
> --- https://julialang.zulipchat.com/#narrow/stream/137791-general/topic/Crazy.20thought.3A.20is.20it.20possible.20to.20merge.20nothing.20and.20missing.3F

---

> Isn't the math different?

I guess you could write `1 ?+ nothing`?  Or, if it's tedious, we can have `@?` like `@.`.  As I commented, I think a more serious problem is the 3VL and ordering.

**Comment by tkf** [comment]
Quoting yet another comment of mine:

> By the way, another consideration of `?` is what should happen with multi-argument functions.  Do we want `f?(a, b; c = c)` to return `missing`/`nothing` if at least one of `a`, `b`, `c` is `missing`/`nothing`?  Or, do we want to have argument-specific syntax like `f(?a, b; c = ?c)` (which just passes `missing`/`nothing` to `f` for `b`)?
>
> --- https://julialang.zulipchat.com/#narrow/stream/137791-general/topic/What.20are.20your.20thoughts.20on.20.3F.20for.20missing.3F/near/203663783

**Comment by DNF2** [comment]
Isn't `missing` a much more specialized and narrow concept than `nothing`? The former is basically just for data scientists, while the latter is useful for anyone who writes code for any purpose. 

I've never used `missing` even once in my life, while `nothing` is the default return value for any function. Isn't `Union{T, Nothing}` the obvious choice?

**Comment by juliohm** [comment]
I think there is no obvious choice because people come from different backgrounds. I personally feel that `?` would be a nice addition for the more "computing-oriented" users, but I acknowledge that many people use Julia for other kinds of applications that do not involve number crunching.

For a vector of objects `xs` it is already very convenient to write `foo.(xs)` and get the evaluation of `foo` on each object. Similarly, it would be very convenient to write `foo.?(xs)` and get the the evaluation on each object that can be possibly missing. The result would be a vector as before but with missing entries. This can clean up a lot of computing-oriented code.

**Comment by clarkevans** [comment]
Let me provide a concrete use case. I'm processing [https://www.hl7.org/](FHIR) data represented as JSON. In the [JSON binding](https://www.hl7.org/fhir/json.html) of FHIR, when data is not provided for a field, it is simply omitted empty key; `null` is not used. When parsed with [JSON.jl](https://github.com/JuliaIO/JSON.jl) we end up with `Dict` structures, and `nothing` never appears. When we process this data, we use `get(dict, field, missing)` since if a value is not provided, the result is missing, and the Clinical Quality Language (CQL) treats missing values with three value logic. 

Hence, the datatypes and functions we've defined for working with FHIR data use quite a bit of `Union{T, Missing}` since that's how the data needs to be processed.  At the top of our files are `MDict = Union{Dict, Missing}`, for example.  In this case, being able to write `T?` as a type, or `function?(args..)` and have them lifted to work with missing values seems very useful.  Having `Union{T, Missing}` everywhere is tedious.

At a more abstract level, I picture `nothing` as the bottom type, in that it signals that there is no value; while, `missing` is the top type, existence of the data is possible but undetermined. Trying to operate on `nothing` does seem like an error, while, having `missing` propagate seems reasonable. In this sense, I think Stefan's [suggestion](https://discourse.julialang.org/t/status-of-question-mark-syntax-for-missing-values/27189/2) makes sense.

**Comment by DilumAluthge** [comment]
Yeah, I think this part of @StefanKarpinski's comment is key:
> However, I’ve come back around to wanting it to mean Union{T, Missing} for one very key reason: we need a way to express lifting of functions to the Union{T, Missing} domain. What I mean by that is that instead of every couple of days getting a new feature request for missing support for the atanh function or whatever, I’d like to just make atanh? lift the atanh function so that it returns missing when any of its arguments are missing. We don’t need this feature for nothing since the whole point of nothing is that it doesn’t lift, whereas the point of missing is to lift.

**Comment by tkf** [comment]
Reposing my comments from https://julialang.zulipchat.com/#narrow/stream/137791-general/topic/Crazy.20thought.3A.20is.20it.20possible.20to.20merge.20nothing.20and.20missing.3F/near/203667599

> I'd replace "lift" with "automatically lift"; then `f?` for `Union{Some{T},Nothing}` suddenly makes sense because it's manual (but with such a terse syntax so that we don't need the automatic one).
>
> Reading this and other Stefan's comments elsewhere, I think his point of view is based on the frustration of so much proliferation of dispatches on `missing` in `Base`.  This is yet another way to put that "automatic lifting" based on dispatch does not scale.  This is not news.  See contextual dispatch systems used in autodiff etc.

> What I call "automatic lifting" is
> 
> ```julia
> julia> 1 + missing
> missing
> ```
> 
> It is automatic from users' perspective.  From the maintainers' perspective, it's manual, tedious, and unscalable.

**Comment by clarkevans** [comment]
> There are several RFCs for adding API based on `Union{Some{T}, Nothing}` (#34821, #33758). For example, it'd be nice `dict?[key]` to return `Union{Some{T}, Nothing}` while it makes less sense to return `Union{T,Missing}` because `dict` may contain `missing`s.

A `dict` may contain `nothing`s too.  I don't see any reason why someone would even store a key in a dictionary if the value is `missing`.  The common access pattern is `get(dict, key, missing)` in my area of work.  Being able to write `dict?[key]` would be quite nice.

**Comment by clarkevans** [comment]
> For a vector of objects `xs` it is already very convenient to write `foo.(xs)` and get the evaluation of `foo` on each object. Similarly, it would be very convenient to write `foo.?(xs)` and get the the evaluation on each object that can be possibly missing. The result would be a vector as before but with missing entries. This can clean up a lot of computing-oriented code.

Wouldn't this be `foo?.(xs)` -- that is, you're broadcasting the lifted function; you'd want a missing returned for each missing input?

**Comment by tkf** [comment]
> A `dict` may contain `nothing`s too.

There is no ambiguity because you'd get `Some(nothing)`.

**Comment by bkamins** [comment]
> It's interesting because I use `Union{T, Nothing}` all the time

@KristofferC - I agree and I also use it often in package code. My comment was about user code (especially in data science related domains). Of course this is opinionated, as I have commented. Though, I think that in package code writing `Unions` is not problematic, where you want to save typing most is interactive sessions. And this is a typical use case for data science work.

Now regarding `fun?` - it is is an orthogonal thing I think. E.g. we currently have `passmissing` in Missings.jl and there was a lot of discussion how the `missing` should be handled there as there is no single best option here if you consider multiple args and multiple kwargs, therefore I would be reluctant to have `fun?` in Base (although if we find that there is one expected pattern that is predominantly common we could adopt it).

**Comment by clarkevans** [comment]
> There is no ambiguity because you'd get `Some(nothing)`.

Sure, but if the value of a dictionary key is `nothing`, why should it return `Some(nothing)`.  This seems just as magical as not being able to distinguish between a key having a value of `missing` and a key for a particular value being absent. 

Anyway, I can see how one might want `dict?[key]` to return `nothing` if the key is absent, and probably it's more conservative approach; since you can't do much with `nothing`.  So, if the type of the Dictionary is `Dict{Any, T}` then the type of the `?[]` operator would be `Union{Nothing, T}` yes? I think that `get(dict, key, missing)` works quite fine in my particular context, I don't need a fancy new syntax for that.  The only odd thing would be if `T?` meant `Union{Missing, T}` then one might expect `dict?[key]` to return `T?`.

**Comment by KristofferC** [comment]
> Sure, but if the value of a dictionary key is nothing, why should it return Some(nothing). 

Because that is how you distinct a value of `nothing` from no value?

**Comment by JeffBezanson** [comment]
I just want to complain about the low quality of this issue (not the discussion, the issue itself). There is no actual proposal, and this seems to be doing no more than forking the discourse thread onto github. A link to discourse is not an acceptable issue description. If I want to read discourse, I'll read discourse.

**Comment by MasonProtter** [comment]
This seems like a very niche thing to have it's own special case in the parser. Wouldn't it make way more sense to just add a method to `|` or `∪` such that it just evaluates to `Union`? This way, we don't privilege `Missing` or `Nothing`, write clearer code and don't add yet another new parser rule. 

You could have
```julia
Base.:(|)(::Type{T}, ::Type{U}) where {T, U} = Union{T, U}

Vector{Int | Nothing} # Array{Union{Int, Nothing}, 1}

struct Foo 
    x::(Int | Missing) # field x must be either Int or Missing 
end
```
or if you prefer
```julia
Base.union(::Type{T}, ::Type{U}) where {T, U} = Union{T, U} # alias ∪

Vector{Int ∪ Missing} # Array{Union{Int, Missing}, 1}

struct Foo 
    x::(Int ∪ Nothing) # field x must be either Int or Nothing 
end
```

**Comment by MasonProtter** [comment]
@JeffBezanson, I hope this isn't taken the wrong way,and I acknowledge that there may be some context I'm missing, but I feel your response here was not very constructive and rather brusque.

I know I'd feel pretty alienated if someone as esteemed in this community as yourself addressed me like that. 

I was going to say that I feel a better response would be to request that @juliohm consult the CONTRIBUTING.MD on how to phrase a feature request that is up to the standards of this repository, but it appears that there isn't really any mention there about language level feature requests, only bug reporting and notes not to make library level feature feature requests, so I opened an issue on that here: https://github.com/JuliaLang/julia/issues/36667.

**Comment by tkf** [comment]
> Wouldn't it make way more sense to just add a method to `|` or `∪` such that of them evaluates to `Union`?

I see that the OP text (rather than the issue title) emphasizes the union type shorthand.  But I fear this is a rather too trivial usage for very valuable ASCII-based syntax.  Rather, I think a better focus here should be the lifting of the function calls like `f?(g?(x, h?(y)) ?+ z)`.  This change would be as fundamental as the dot call syntax we use everywhere.

**Comment by MasonProtter** [comment]
I think that even if this weren't a very niche thing which broadcast isn't, I'd still be against it just because of the huge amount of ambiguity it'd add to the already confusing ternary operator syntax. If anything, we should be thinking of how we can have *less* special syntactic forms in the parser, not more. 

I also think the use-case you show above is just better handled by a macro anyways:
```julia
julia> using MacroTools, Missings

julia> @eval macro $(Symbol("?"))(expr)
           (esc ∘ MacroTools.postwalk)(expr) do ex
               if ex isa Expr && ex.head == :call
                   ex.args[1] = :($(passmissing)($(ex.args[1])))
               end
               ex
           end
       end
@? (macro with 1 method)

julia> f(x::Int) = x + 1
f (generic function with 1 method)

julia> @? sin(f(missing) * 2)
missing
```

**Comment by DilumAluthge** [comment]
> I think it could be worthwhile to have a JuliaCon BoF topic to hammer this discussion out, since we can't rely on hallway conversations this year! And syntax would be really nice for this.

It seems (from this issue discussion, as well as discussions on Discourse and Zulip) that this is a sufficiently complicated topic to justify a BoF at JuliaCon as @vtjnash suggested

**Comment by tkf** [comment]
There are reasons why it's nice to have native dot calls and not just `@.`.  It is very useful to be explicit and concise about what you lift.  Lift-everything macro is indeed possible to implement although not trivial (I'm actually writing Maybe.jl that does that).  Furthermore, it may be useful to lift on arguments like `f(?a, b, ?c, d)` rather than the entire function `f?(a, Some(b), c, Some(d))`.  It would be more syntax-heavy to do this in a macro.  Another reason in favor of the native syntax is that macros do not compose well.

**Comment by piever** [comment]
> Rather, I think a better focus here should be the lifting of the function calls like f?(g?(x, h?(y)) ?+ z). This change would be as fundamental as the dot call syntax we use everywhere.

I also feel that this is the key point. In my view, the decision is whether we want `f?` to be syntax to grant composability of functions that can return either one or zero things, or to be syntax to handle missing data.

In particular, I'm not sure I agree with this comment:

> We don’t need this feature for `nothing` since the whole point of `nothing` is that it doesn’t lift, whereas the point of `missing` is to lift.

Because to me the point of `nothing` is that it does not lift _implicitly_, but lifting over it explicitly makes sense (for example, the  `Option` type in rust has a `map` method for this).

My two questions are:

1. Does there exist a consistent definition of `f?` that would remove the need of all the special casing of `missing` in Base?
2. If 1. has a positive answer, what is the advantage of separating `missing` and `nothing`, given that propagation would be done via `?` rather than by whitelisting a set of functions?

**Comment by clarkevans** [comment]
I think this discussion may wish to review if this special `?` syntax should apply to `Missing` or `Nothing` (presumably not both).  Generally it has not been a burden for me to write  `OptString = Union{String, Missing}` and use `OptString` everywhere. There is only minor ergonomic improvement writing `String?` for use in 3VL (aka `Missing` data).  I'm not sure a generic lifting semantics for functions with potentially-missing arguments is even possible, moreover, I'm not sure that we want users of functions to have to know when to use `?` and when to not use it... stuff like this can get confusing.  Perhaps it's better if functions that handle `Missing` arguments do so explicitly.  Moreover, I don't mind `get(mydict, "mykey", missing)`, it's rather clear what's going on and if `mydict["mykey"] == missing` then I'm alright with the ambiguity (was it actually a `missing` value or was the key missing), in most cases, this ambiguity is a feature.  Hence, for casual use, where 3VL logic is used, I'm not sure if additional automatic lifting or syntax sugar will be a net positive; things already work relatively well.

On the other hand, `Nothing` is the more explicit cousin and it's not easy to grok or use -- perhaps it deserves to have an special syntax and explicit lifting semantics?  Even those that are supporting data scientists end up putting on the system programmer hat where 3VL is not what we wish. In these cases, a nicer syntax might encourage proper use of `Nothing` (a value does not exist) over `Missing` (the value is unknown) as the default when "nothingness" is needed. This seems to be the direction that @tkf is suggesting.

Regardless, Jeff has a great point about venue; I suppose we should be using Discourse? (since Slack history disappears)  Could there be a "Feature Request" section in Discourse, perhaps under "Internals & Design" and shuffle discussion there till enough consensus emerges that we have a feature set suitable for review?

**Comment by tkf** [comment]
> 1. Does there exist a consistent definition of `f?` that would remove the need of all the special casing of `missing` in Base?

@piever I think the answer is "no."

For concrete definition of "lifting",  this function may be a nice toy example:

```julia
return1(_) = 1
```

This function picks the right argument.  The question is what `return1?(x)` should do.

In the `Union{Some{T},Nothing}` world, it's pretty straightforward:

```julia
return1?(Some(x)) === Some(1)
return1?(nothing) === nothing
```

<details>
 <summary>Note: equivalent code in Haskell's maybe monad</summary>

```hs
Prelude> Just 1 >>= return . const 1
Just 1
Prelude> Nothing >>= return . const 1
Nothing
```

</details>

It is also straightforward to define `f?` for `Union{Some{T},Nothing}` generally:

<details>
 <summary>An implementation of <code>f?</code></summary>

```julia
f? = (args...; kwargs...) -> maybecall(f, args...; kwargs...)

function maybecall(f, args...; kwargs...)
    any(isnothing, args) && return nothing
    any(isnothing, Tuple(kwargs.data)) && return nothing
    y = f(_somethings(args)...; _somethings(kwargs.data)...)
    return y === nothing ? y : Some(something(y))
end

_somethings(xs::Tuple) = map(something, xs)
_somethings(xs::NamedTuple{names}) where {names} =
    NamedTuple{names}(_somethings(Tuple(xs)))
```

</details>

In the `Union{T,Missing}` world, what `return1?(x)` should do?  I'm guessing

```julia
return1?(x) === 1
return1?(missing) === 1
```

The main difference to `Union{Some{T},Nothing}` is `return1?(missing) === 1` (@juliohm @bkamins @clarkevans @piever please point it out if this definition doesn't make sense).

So, I think using `f?` for `Union{T,Missing}` has a large conceptual issue.  This is because it's hard to define "lifting" mechanistically (like `return1?(missing, y)`).  What `f?` should do depends on the semantics of the code.

This is why I think `f?` should be used for `Union{Some{T},Nothing}` and not `Union{T,Missing}`.

> I'm not sure that we want users of functions to have to know when to use `?` and when to not use it

@clarkevans I think it's rather dangerous that users need to know which functions _implicitly_ special-case `missing` in "non generic" way even they are not seeing special syntax.  Due to this, it is invalid to use

```julia
if x == y
    ...
end
```

in a generic code because `if` requires a `Bool` and `x == y` could evaluate to `missing`.

> My question about this is nesting, on the Nth level, will I have `Some(Some(...Some(nothing)...))` ?

I'm not sure what you mean.  The direct answer is yes.  This an important feature.  However, I'm suspecting that this could be coming from a misunderstanding of how lifting works for `Union{Some{T},Nothing}`.  Note that `Some?(Some?(Some?(Some(nothing))))` is `Some(nothing)`.

**Comment by tkf** [comment]
FYI here is my POC for `@?`: https://github.com/tkf/Maybe.jl

**Comment by nalimilan** [comment]
> My two questions are:
> 
>     1. Does there exist a consistent definition of `f?` that would remove the need of all the special casing of `missing` in Base?
> 
>     2. If 1. has a positive answer, what is the advantage of separating `missing` and `nothing`, given that propagation would be done via `?` rather than by whitelisting a set of functions?


@piever One difficulty is that the three-valued logic implemented by logical operators differs from standard lifting semantics. So lifting cannot replace all operations defined on `missing`. Though we could define these as special cases: that shouldn't be too confusing as the list of logical operators is very short.


> In the `Union{T,Missing}` world, what `return1?(x)` should do? I'm guessing
> 
> ```julia
> return1?(x) === 1
> return1?(missing) === 1
> ```
> 
> The main difference to `Union{Some{T},Nothing}` is `return1?(missing) === 1` (@juliohm @bkamins @clarkevans @piever please point it out if this definition doesn't make sense).
> 
> So, I think using `f?` for `Union{T,Missing}` has a large conceptual issue. This is because it's hard to define "lifting" mechanistically (like `return1?(missing, y)`). What `f?` should do depends on the semantics of the code.

@tkf I don't get why you would want `return1?(missing) === 1`. The point of lifting is to return `missing` when one of the inputs is `missing` -- just like for `nothing`. See what `passmissing(f)` does in Missings.jl.

**Comment by clarkevans** [comment]
It seems that this is a popular topic.  There is a [blog post[(https://dart.dev/null-safety/understanding-null-safety) on `String?` in Dart.  The idea that the compiler could know about potentially `missing` or `nothing` values, and help the developer avoid mistakes is quite attractive.

**Comment by tkf** [comment]
@nalimilan I know what `passmissing` does.  I was just pointing out that you can't get something like the 3VL from the maybe monad.

**Comment by ssagaert** [comment]
"Maybe T? for missing and T?? for nothing. "

If you want shortcuts for both then I'd switch it around because T? means T | Null  or Nullable in languages that support that notation (Kotlin, C#,...) and that maps to Union{T,Nothing}.

**Comment by bkamins** [comment]
The issue is that AFAICT do not have `missing` as a part of a language.

**Comment by juliohm** [comment]
Did we reach a conclusion regarding this FR?

**Comment by bkamins** [comment]
No, AFAICT the discussion is stalled. Still - I personally would support parsing `Int?` as `Union{Int, Missing}`.

**Comment by StefanKarpinski** [comment]
I've come to regret including Missing in Base Julia and am therefore reluctant to give it dedicated syntax. If the syntax could be used for different things and loading `Missings` defined it, that would be ok. However, from my admittedly biased experience I think using `T?` for `Union{T, Nothing}` seems like a better use of the syntax.

**Comment by ssagaert** [comment]
> I think using T? for Union{T, Nothing} seems like a better use of the syntax.
> 

That was my proposal. This is also inline with what other languages do (C#, Kotlin,…)

**Comment by jariji** [comment]
Imho `Union{T, Nothing}` is one of those often too-convenient features that lures people into unsafe usage where they have no proof ruling out `Nothing <: T` and hence they accidentally merge the failure case with the success-with-value-`nothing` case. 

Eg if `findfirst` is used inside a library where `T` is generic and set by the (external) user the library will do the wrong thing.
```jl
julia> findfirst(==(1), Dict(nothing=>10, 2=>20))
# incorrectly assume it wasn't found
```

Using `Some` to distinguish success from failure is preferable for generic behavior.


This issue is discussed more in https://github.com/JuliaLang/julia/issues/29565 and by Takafumi in this very thread above. https://github.com/JuliaLang/julia/issues/36628#issuecomment-659146056

**Comment by mbauman** [comment]
Must it be one meaning or the other?  Why can't `?` just be a free-for-all postfix operator?  Doing `using Missings: ?` gives you the `passmissing`/`Union{T, Missing}` behavior, whereas `using Base: ?` could be the `nothing` thing.  The Base meaning could be public but not exported.

**Comment by clarkevans** [comment]
> I've come to regret including Missing in Base Julia and am therefore reluctant to give it dedicated syntax. If the syntax could be used for different things and loading `Missings` defined it, that would be ok. However, from my admittedly biased experience I think using `T?` for `Union{T, Nothing}` seems like a better use of the syntax.

We've been working with SQL databases often. It's great to have `Missing` data type; I think including it in Base was the correct decision, otherwise every library would be representing `NULL` values in their own way. Even so, I agree with the conclusion: we don't need any special syntax for working with `Missing`. I cannot remember when we've needed to have types in Julia that worked with `Missing`, except at the library level, where it's perfectly fine to write `Union{String, Missing}`.

Conversely, having better conventions and tools for working with `Nothing` and seems useful. I often want to have arguments default to `nothing` and `something()` is useful. Being able to declare `String?` seems helpful. It would also make it more clear that `Nothing` is the Julia approach and that `Missing` is more for handling data form external systems that use 3-valued logic.

**Comment by adienes** [comment]
Kotlin's "elvis operator" is kind of snazzy, where `x ?: y`  means `isnothing(x) ? y : x`

e.g. `findfirst(==(1), [2,3,4]) ?: default_value`

**Comment by jariji** [comment]
The elvis operator is too late: once `findfirst` does its dark magic we've already lost track of which branch we're on. 

(If we're gonna keep going on this point let's do it in https://github.com/JuliaLang/julia/issues/29565)

**Comment by Keno** [comment]
There has been discussion of this issue again, related to a different use of postfix-? in #60344. However, I feel that there is a bit of a lack of motivating examples in this particular issues (a bit of an echo of Jeff's complaint above). Of course part of that is there isn't actually a proper proposal on the table in this issue, but my question is this: Can somebody provide an example where you'd write a function that needs to do a bunch of lifted operations in a row, such that a dedicated postfix-? (as opposed to a generic `lift`) would provide significant readability advantages. To be clear, I can see the monadic version being very useful (imagining in this code snippet that `?` is monadic propagation), because writing the control flow is annoying.
```
function find_person(uuid, age)::Union{Person, Missing}
    ps = people()? # equivalent to ps = people(); ps === missing && return missing
    p = findfirst(p->p.uuid == uuid, ps)?
    p.age? == age || return missing
    return p
end
```
but I don't know that I can see a similarly good argument for lifting, so if somebody has it I'd like to see it.

(This is a genuine request, I don't do very much data science so I'm not familiar with the code patterns).

**Comment by Keno** [comment]
Still looking for motivating examples.

**Comment by Keno** [comment]
> Still looking for motivating examples.

Call the third for motivating examples.

**Comment by bkamins** [comment]
I think the main use case I see is to have:

> type `T?` to indicate `Union{T,Missing}`

(as in original comment in this issue)

The usage would be twofold:
1. Simplifying writing and reading function signatures and composite and parametric type specifications
2. Also when printed `Union{T,Missing}` could be replaced by `T?` which would make it nicer to read

Lifting is not so crucial in my opinion.

**Comment by quinnj** [comment]
I agree that lifting doesn't seem like the most valuable use of this syntax. I think my bias lately has 2 main pulls:
* having to read/write occasional javascript
* doing more "generic software engineering" work vs. data munging

Hence, my preference would be to match the javascript use of [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) (explicit `nothing` propagation) and [null coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) (basically `nothing || foo`) specifically with Julia's `nothing` (instead of `missing`). It would make a lot of generic domain-object struct usage much more ergonomic.

**Comment by pdeffebach** [comment]
@Keno @nalimilan 

Here is some motivation for a simplified syntax. Lots of very generic data cleaning operations require `passmissing` and `skipmissing`, to the extent that a nicer syntax would be welcome. 

I don't do a lot of data cleaning in Julia right now (partially because missings very annoying, but also because I'm working on projects with teams that use R and Python). 

But I just found a messy [dataset](https://www.kaggle.com/datasets/ahmedmohamed2003/cafe-sales-dirty-data-for-cleaning-training) on Kaggle and cleaned it the way I would any other dataset. Hopefully the cleaning code shows  how much extra syntax is needed to handle missing values in Julia. 

1. Control flow. Using `if...else` breaks with missing values. To get that to work, you have to wrap the operation in a function and use `passmissing`, which is a pain. 
2. Parsing. Parsing functions don't accept missing values, so you have to use `passmissing`. This is especially annoying because in theory, `parse` should be able to handle diverse inputs (or tryparse, maybe). Once again you have to use `passmissing`. 
3. Functions with strings. While mathematical functions propagate `missing` values, string functions do not. This also creates lots of `passmissing` wrappers. 
5. Means with missing values. To take the mean of a vector with missing values, you have to use `skipmissing`. If you take a lot of means, this gets tired quickly. 
6. Working with two vectors simultaneously which have missing values. You have to make sure you get matching non-missing observations and then collect them together. 
7. Graphing. Makie doesn't work well with missing values. This is not totally Base's concern, but maybe with some standardized syntax we can see stronger support. 

* At no point in this example did I feel the need to do `T?` to represent `Union{T, Missing}`. That feels like a pain point for developers, but not necessarily for users. 
* It seems very unlikely that `?` will be used for `skipmissing`, or other automatic handling of missing values in collections. But its still annoying to type `skipmissing` too much. 

<details>
<summary>Cleaning code</summary>

```julia
# Cleaning messy cafe sales

using Dates

using CSV

using DataFramesMeta, AddToField, MissingsAsFalse, Missings

using Statistics, StatsBase

using GLMakie, AlgebraOfGraphics, PrettyTables

# Import the raw data

df_raw = @chain begin
	"data/dirty_cafe_sales.csv"
	CSV.read(DataFrame)
end


# Clean the column names
df = @chain df_raw begin 
	rename(x -> replace(lowercase(x), " " => "_"), _)
end

# Replace UNKNOWN and ERROR with missing 
# This does not work well with missing. It errors in the comparison to 
# "UNKNOWN" and "MISSING" ? and || The way I get around this is by 
# making a function and calling passmissing. But I would have 
# rather written it inline. 
function fix_unknown_error(x)
	(x == "UNKNOWN" || x == "ERROR") ? missing : x
end

df = @chain df begin
	@rtransform begin 
	 	:item = passmissing(fix_unknown_error)(:item)
	 	:quantity = passmissing(fix_unknown_error)(:quantity)
	 	:price_per_unit = passmissing(fix_unknown_error)(:price_per_unit)
	 	:total_spent = passmissing(fix_unknown_error)(:total_spent)
	 	:payment_method = passmissing(fix_unknown_error)(:payment_method)
	 	:location = passmissing(fix_unknown_error)(:location)
	 	:transaction_date = passmissing(fix_unknown_error)(:transaction_date)
	end
end

# Parse the string columns as numbers 
# Another wrinkle here. The parse functions
# don't allow missing values. You need to 
# call passmissing here 
df = @chain df begin 
	@rtransform begin 
		:quantity = passmissing(parse)(Int, :quantity)
		:price_per_unit = passmissing(parse)(Float64, :price_per_unit)
		:total_spent = passmissing(parse)(Float64, :total_spent)
		:transaction_date = passmissing(Dates.Date)(:transaction_date, "yyy-mm-dd")
	end
end

# Make the item column lower-case so its easier to work with programatically
# Functions that work with strings don't allow missing values.
# So we need to wrap in passmissing here. 
df = @chain df begin 
	@rtransform :item = passmissing(lowercase)(:item)
end

# Now let's make a table of the average total spent per item
# Notice that we have to use "skipmissing" here to 
# handle missings. 
# Obviously this will happen a lot, so a syntax to drop
# missings easier is useful. 
t = @chain df begin 
	@by :item begin 
		:avg_spent = mean(skipmissing(:total_spent))
	end
end

# Get the correlation between total spent and 
# quantity
# This is hard, because you have to use skipmissings (note the s)
# to get jointly shared observations, but then also 
# collect them
cor_q_spend = @with df begin 
	s_q, s_t = skipmissings(:quantity, :total_spent)
	s_q = collect(s_q)
	s_t = collect(s_t)
	cor(s_q, s_t)
end

# Plot a histogram of total spent by 
# payment method
# Another problem: You can't use missing values 
# here. I think the missing values are 
# automatically promoted to NaN. Need
# to drop missing, I think?
p = @chain df begin 
	dropmissing([:total_spent, :payment_method])
	data(_) * 
		mapping(:total_spent, color = :payment_method) * 
		visual(Density, alpha = 0.2, strokewidth = 2)
	draw()
end
```

</details>

**Comment by adienes** [comment]
personally speaking I would much rather see `?` associated with a rust inspired `Some` - or - `nothing` like api rather than taken for `missing` handling.

I already feel like `missing` handling and passthrough is over-proliferated in too many apis and should really only exist in statistical or data handling contexts.
